lib/azure/core/http/http_request.rb

```diff
@@ -0,0 +1,175 @@
+        # Public: Applies a HttpFilter to the HTTP Pipeline
+        #
+        # filter - Any object that responds to .call(req, _next) and
+        #          returns a HttpResponse eg. HttpFilter, Proc,
+        #          lambda, etc. (optional)
+        #
+        # &block - An inline block may be used instead of a filter
+        #
+        #          example:
+        #
+        #             request.with_filter do |req, _next|
+        #               _next.call
+        #             end
+        #
+        # NOTE:
+        #
+        # The code block provided must call _next or the filter pipeline
+        # will not complete and the HTTP request will never execute
+        #
+        def with_filter(filter=nil, &block)
+          filter = filter || block
+          if filter
+            old_impl = self._method(:call)
+
+            # support 1.8.7 (define_singleton_method doesn't exist until 1.9.1)
+            new_impl = Proc.new do
+              filter.call(self, old_impl)
+            end
+            k = class << self;
+              self;
+            end
+            if k.method_defined? :define_singleton_method
+              self.define_singleton_method(:call, new_impl)
+            else
+              k.send(:define_method, :call, new_impl)
+            end
+          end
+        end
+
+        # Build a default headers Hash
+        def default_headers(current_time)
+          {}.tap do |def_headers|
+            def_headers['User-Agent'] = Azure::Core::Default::USER_AGENT
+            def_headers['x-ms-date'] = current_time
+            def_headers['x-ms-version'] = '2014-02-14'
+            def_headers['DataServiceVersion'] = '1.0;NetFx'
+            def_headers['MaxDataServiceVersion'] = '2.0;NetFx'
+            def_headers['Content-Type'] = 'application/atom+xml; charset=utf-8'
+          end
+        end
+
+        def http_setup
+          http = @client.agents(uri)
+
+          unless headers.nil?
+            keep_alive = headers['Keep-Alive'] || headers['keep-alive']
+            http.read_timeout = keep_alive.split('=').last.to_i unless keep_alive.nil?
+          end
+
+          http
+        end
+
+        def body=(body)
+          @body = body
+          apply_body_headers
+        end
+
+        # Sends request to HTTP server and returns a HttpResponse
+        #
+        # @return [HttpResponse]
+        def call
+          conn = http_setup
+          res = conn.run_request(method.to_sym, uri, nil, nil) do |req|
+            req.body = body if body
+            req.headers = headers if headers
+          end
+
+          response = HttpResponse.new(res)
+          response.uri = uri
+          raise response.error unless response.success?
+          response
+        end
+
+        private
+
+        def apply_body_headers
+          if body
+            if IO === body
+              headers['Content-Length'] = body.size.to_s
+              headers['Content-MD5'] = Digest::MD5.file(body.path).base64digest
```

Comment:
We may add: unless headers['Content-MD5']. because we may allow caller to set the MD5 by himself.